### PR TITLE
Implemented #168 and #171

### PR DESCRIPTION
**Pull request by gorance-teletrader**
This is updated version of app, this features are added.
#168 Add protocol to open app from browser
Example 1:                              jitsi://test
Example 2 (custom server):    jitsi://test/<some_server_uri>

#171 Disable window always on top
Create config
After changing the setting, reload of app is needed to applay changes
Note: due to lack of time, **auto reload** and **sync between stores** (browser and local file) are not implemented _yet_ :)

I tried to follow style of writing, but if something is not ok i will changed it.

One note, this is only tested on Windows 10
and could have bugs on different platforms

### CODE REVIEW

File: main.js [hunk 1]
```diff
@@ -211,3 +286,66 @@ app.on('window-all-closed', () => {
         app.quit();
     }
 });
+
+// remove so we can register each time as we run the app.
+app.removeAsDefaultProtocolClient(PROTOCOL_PREFIX);
+
+// If we are running a non-packaged version of the app && on windows
+if (process.env.NODE_ENV === 'development' && process.platform === 'win32') {
```
Comment (akshitkrnagpal) [inline main.js]: I think you can use `isDev` here.

**Comment by saghul** [comment]
> Hi @gorance-teletrader
> 
> Thank for contributing. 😄
> I have left some comments on the PR.
> I also think this should be split into different PRs. (@saghul can you confirm 😄)

I susbribe to everything @akshitkrnagpal has said here :-) Plus an additional thank you for contributing!

**Comment by gorance-teletrader** [comment]
Thanks for fast replay 👍 
i have fixed "problems" and ones that i have find while coding this

I don't have time to split PRs, but i agree that is better (i should think about it when i started)

File: app/features/config/index.js [hunk 2]
```diff
@@ -28,5 +28,26 @@ export default {
...
      * The URL of Terms and Conditions Page.
      */
-    termsAndConditionsURL: 'https://jitsi.org/meet/terms'
+    termsAndConditionsURL: 'https://jitsi.org/meet/terms',
+
+    /**
+     * Some of default values for application.
+     */
+    defaults: {
+        windowAlwaysOnTop: true
```
Comment (akshitkrnagpal) [inline app/features/config/index.js]: I think `alwaysOnTopWindowEnabled` would describe this flag better. 😄

File: package-lock.json [hunk 3]
```diff
@@ -5944,8 +5958,17 @@
       "requires": {
         "nan": "^2.2.1",
         "postis": "^2.2.0",
-        "prebuild-install": "^2.4.1",
-        "robotjs": "github:jitsi/robotjs#5cc469f655669e728b61271c2c57ce97d62976fd"
+        "prebuild-install": "^2.4.1"
+      },
+      "dependencies": {
+        "robotjs": {
```
Comment (akshitkrnagpal) [inline package-lock.json]: I think the Travis build might be breaking because if this `package-lock.json` change. Since `package.json` does not have any change the update in lock file might be due to version mismatch of `npm` or `node`.

Travis uses node `v8.16.0` and `npm 6.4.1`.

Can you tell the version of node and npm you are using?

Comment (gorance-teletrader) [reply package-lock.json]: Hi, i'm using 
Node: 10.15.3 
npm:   6.4.1 (same)

File: app/features/app/components/App.js [hunk 4]
```diff
@@ -14,18 +16,58 @@ import { Welcome } from '../../welcome';
...
+     * Handler when main proccess contact us.
+     *
+     * @param {Object} event - Message event triggered by .
+     * @param {Object} arg - Object with two props same as Conference object.
+     *
+     * @returns {void}
+     */
+    _listenOnProtocolMessages = (event, arg) => {
+        // change route when we are notified
+        // eslint-disable-next-line no-invalid-this
```
[the remark this answers is not in the document]
Comment (gorance-teletrader) [reply app/features/app/components/App.js]: Sorry for this :)
I'm used to my react setup

File: app/features/config/index.js [hunk 5]
```diff
@@ -28,5 +28,26 @@ export default {
...
+    /**
+     * Some of default values for application.
+     */
+    defaults: {
+        alwaysOnTopWindowEnabled: true
+    },
+
+    /**
+     * All needed configurations for storage in our application
+     * This is including electron-store and redux-persist
```
Comment (saghul) [inline app/features/config/index.js]: I don't want to have 2 ways to persist storage. Let's exist with the existing redux-persist please.

File: app/features/config/index.js [hunk 6]
```diff
@@ -28,5 +28,26 @@ export default {
...
+     */
+    storage: {
```
Comment (saghul) [inline app/features/config/index.js]: This doesn't need to be configurable, please undo this change.

File: app/features/settings/components/WindowAlwaysOnTopToggle.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    /**
+     * Initializes a new {@code WindowAlwaysOnTopToggle} instance.
+     *
+     * @inheritdoc
+     */
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            windowAlwaysOnTop: true
```
Comment (saghul) [inline app/features/settings/components/WindowAlwaysOnTopToggle.js]: You don't need to store state here, just use the prop value.

File: app/features/config/index.js (hunk 5 above)
Comment (gorance-teletrader) [reply app/features/config/index.js]: I don't understand comment, you need to use electron-store to access redux-persist data from main.js
And redux-persists uses electron-store to read/write data on user device

File: app/features/config/index.js (hunk 6 above)
Comment (gorance-teletrader) [reply app/features/config/index.js]: Later, if you want to change **rootKey** in redux persists (change only in config) your app is still working, because main.js uses this config to access stored data

File: app/features/settings/components/WindowAlwaysOnTopToggle.js (hunk 7 above)
Comment (gorance-teletrader) [reply app/features/settings/components/WindowAlwaysOnTopToggle.js]: i will, but i was following style of [settings components](https://github.com/jitsi/jitsi-meet-electron/blob/master/app/features/settings/components/StartMutedToggles.js#L55)

Comment (akshitkrnagpal) [reply app/features/settings/components/WindowAlwaysOnTopToggle.js]: @saghul 
I think this was done because of `isDefaultChecked` prop. Don't remember what problem was it causing.

File: main.js [hunk 8]
```diff
@@ -141,23 +161,144 @@ function createJitsiMeetWindow() {
...
+/**
+ *
+ * Get data from store, and parse what is needed
+ * So other functions can access data as objects
+ *
+ * There is currently limitation
+ * persist data is typeof string
+ * so we cannot just do `store.get('a.b.c');
+ */
+function parsePersistStore() {
```
[the remark this answers is not in the document]
Comment (gorance-teletrader) [reply main.js]: Its my pleasure, i was thinking about that
but i need to find more free time to test this one

File: app/features/config/index.js (hunk 5 above)
Comment (saghul) [reply app/features/config/index.js]: Oh, I'll check it again then.

File: app/features/config/index.js (hunk 6 above)
Comment (saghul) [reply app/features/config/index.js]: We don't want that to be configurable.

File: app/features/settings/components/WindowAlwaysOnTopToggle.js (hunk 7 above)
Comment (saghul) [reply app/features/settings/components/WindowAlwaysOnTopToggle.js]: Oh, right!

File: main.js [hunk 9]
```diff
@@ -141,23 +161,144 @@ function createJitsiMeetWindow() {
...
+    // This is required for some features
+    // Google Sign In - Start Youtube Live Stream, Settings -> Calendar
+    // Microsoft Sign In - Settings -> Calendar
+    // Dropbox Sign In - Start Recording Feature.
+    //
     initPopupsConfigurationMain(mainWindow);
-    setupAlwaysOnTopMain(mainWindow);
+
+    if (isWindowAlwaysOnTop) {
+        setupAlwaysOnTopMain(mainWindow);
```
Comment (akshitkrnagpal) [inline main.js]: Hey @gorance-teletrader 
Hey @gorance-teletrader 

I just took another look and I found that `setupAlwaysOnTopRender` is still being executed in Conference component. Can you make that prop dependent? 😄 

Also, please check what is the effect of keeping `setupAlwaysOnTopMain` and just not doing `setupAlwaysOnTopRender` if the prop it false. 

It might clean up main a lot. 😄

File: package.json [hunk 10]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "jitsi-meet-electron",
-  "version": "1.1.1",
+  "version": "1.2.0",
```
[the remark this answers is not in the document]
Comment (akshitkrnagpal) [reply package.json]: On the second thought, you can remove the version bump from this. We might not release a new version immediately after this PR.

File: app/features/config/index.js (hunk 6 above)
Comment (wilhelmy) [reply app/features/config/index.js]: I assume this is about the visibility of the little video window... A friend of mine and I are having problems with that window popping up whereever we go, and we both find it kind of intrusive about using the electron client. I think he fixed it by some workaround that closes it whenever it appears, I just wanted to mention that this actually causes usability problems for some people.

**Comment by kempu** [comment]
Thank you @gorance-teletrader for this awesome code, saved me days of research :)
To get this functionality released, I have updated the code to the latest app version and split these functionalities into separate pull requests (#263 and #264) based on maintainers request.

**Comment by saghul** [comment]
Continues in #263 and #264
